src/nfa/operations.cc

```diff
@@ -114,6 +114,364 @@ namespace {
+                    continue;
+
+                if (macrostate_vec[j].IsSubsetOf(macrostate_vec[i])) {           // found covering state
+                    covering_set.insert(macrostate_vec[j]);               // is not covered
+                    covering_indexes.push_back(j);
+                }
+            }
+
+            if (covering_set == macrostate_vec[i]) {
+                for (unsigned long k = 0; k < covering_indexes.size()-1; k++) {      // check resurse coverability
+                    if (macrostate_vec[covering_indexes[k]].size() == 1)            // end on single-sized
+                        break;
+
+                    if (visited[covering_indexes[k]])                               // already processed
+                        continue;
+
+                    visited[covering_indexes[k]] = true;
+
+                    residual_recurse_coverable(macrostate_vec, covering_indexes, covered, visited, k, subset_map, result);
+                }
+
+                covering_set.clear();                 // clear variable to store only needed macrostates
+                for (auto index : covering_indexes) {
+                    if (covered[index] == 0) {
+                        auto macrostate_ptr = subset_map->find(macrostate_vec[index]);
+                        if (macrostate_ptr == subset_map->end())        // should never happen
+                             throw std::runtime_error(std::to_string(__func__) + " couldn't find expected element in a map.");
+
+                        covering_set.insert(macrostate_ptr->second);
+                    }
+                }
+
+                remove_covered_state(covering_set, subset_map->find(macrostate_vec[i])->second, result);
+                covered[i] = true;
+            }
+        }
+
+        delete subset_map;  // clean up
+        return result;
+    }
+
+    Nfa reduce_size_by_residual(const Nfa& aut, StateRenaming &state_renaming, const std::string& type, const std::string& direction){
+        Nfa back_determinized = aut;
+        Nfa result;
+
+        if (direction != "forward" && direction != "backward"){
+            throw std::runtime_error(std::to_string(__func__) +
+                                 " received an unknown value of the \"direction\" key: " + direction);
+        }
+
+        // forward canonical residual automaton is firstly backward determinized and
+        // then the residual construction is done forward, for backward residual automaton
+        // is it the opposite, so the automaton is reverted once more before and after
+        // construction, however the first two reversion negate each other out
+        if (direction == "forward")
+            back_determinized = revert(back_determinized);
+        back_determinized = revert(determinize(back_determinized));          // backward deteminization
+
+        // not relly sure how to handle state_renaming
+        (void) state_renaming;
```

Comment:
It does not make sense in this context, as you pointed out. I agree that this is the best one can do now. When the interface gets changes, this should be taken into consideration as well.